### PR TITLE
[ONNX] Introduce FxFrontend for modularization and experimentations

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96781
* __->__ #95651
* #96935
* #95935

- [ ] Return guards from FxFrontend.
- [ ] IOWrangler.
- [ ] Move unstable fx frontends to testing.

### CODE REVIEW

**Comment by BowenBao** [comment]
CI Error on xdoctest due to `frontend.py` has a few docstring that contains example python script. The error is from the import dependency on 'onnxscript' from `frontend.py`. Need to figure out a way to skip the file for xdoctest. Or conditionally testing when onnxscript is installed.

**Comment by BowenBao** [comment]
> CI Error on xdoctest because `frontend.py` was included in the test due to having a few docstring that contains example python script. The error is from the import dependency on 'onnxscript' from `frontend.py`. Need to figure out a way to skip the file for xdoctest. Or conditionally testing when onnxscript is installed.
> 
> The error is not from the docstring itself, so the skip inside docstring does not help.

https://github.com/pytorch/pytorch/pull/95686 should relax it.

File: torch/onnx/_internal/fx/frontend.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        """Initialize the frontend with config arguments."""
+        ...
+
+    @abc.abstractmethod
+    @_beartype.beartype
+    def _inner_call(self, model: Callable, *args, **kwargs) -> torch.fx.GraphModule:
+        ...
+
+    @_beartype.beartype
+    def __call__(self, model: Callable, *args, **kwargs) -> torch.fx.GraphModule:
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/frontend.py:39]: Perhaps `def trace` is better?

File: torch/onnx/_internal/fx/frontend.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        self.captured_graph_count = 0
+
+    def compile(self, graph_module: "torch.fx.GraphModule", _):
+        assert self.captured_graph_count == 0
+        self.captured_graph = graph_module
+        self.captured_graph_count += 1
+        return graph_module
+
+
+class AOTAutogradFrontend(FxFrontend):
```
Comment (wschin) [inline torch/onnx/_internal/fx/frontend.py:206]:
```suggestion
class DynamoOptimize(FxFrontend):
```

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+import torch
+from torch import nn
+from torch.onnx._internal.fx import frontend
+from torch.testing._internal import common_device_type, common_modules, common_utils
+from torch.utils import _pytree as pytree
+
+_FrontendType = Union[frontend.FxFrontend, frontend.FxFrontendUnpackKwargs]
+FX_FRONTENDS: Sequence[_FrontendType] = [
+    # TODO(bowbao): skip aotautograd, because it alters graph inputs/outputs.
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:20]: Can you test this a simple nn.Module or a 1-op function so that aotautograd doesn't alter the function? This makes me  feel we need to have the ability to select what frontend to test for which model.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+):
+    make_input = functools.partial(
+        common_modules.make_tensor,
+        device=device,
+        dtype=dtype,
+        requires_grad=requires_grad,
+    )
+    module_inputs = [
+        common_modules.ModuleInput(
+            constructor_input=common_modules.FunctionInput(),
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:60]: What is `constructor_input` input? Why is `common_modules.FunctionInput()` assigned to it? Does it mean a `None`?

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            MODULES_DB, train_eval_mode=common_modules.TrainEvalMode.eval_only
+        )
+        def _output_match_base(
+            self: common_utils.TestCase,
+            device: str,
+            dtype: torch.dtype,
+            module_info: common_modules.ModuleInfo,
+            training: common_modules.TrainEvalMode,
+        ):
+            """Base test method for testing each fx frontend, used by instantiate_device_type_tests."""
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:87]: Need more documentation in this file.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            # device is provided by instantiate_device_type_tests, but we only want to run in cpu.
+            assert device == "cpu"
+
+            module_cls = module_info.module_cls
+            module_inputs = module_info.module_inputs_func(
+                module_info,
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:93]: `module_inputs_func` is a class method of `module_info`. Why is it called with `module_info`?

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                    sample_num=i,
+                    fx_frontend=fx_frontend.name,
+                ):
+                    init_args, init_kwargs = (
+                        module_input.constructor_input.args,
+                        module_input.constructor_input.kwargs,
+                    )
+                    model = module_cls(*init_args, **init_kwargs)
+                    model.to(device).to(dtype)
+                    model.eval()
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:112]: Probably need a parameter to control if this should be .train or .eval.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+                    args, kwargs = (
+                        module_input.forward_input.args,
+                        module_input.forward_input.kwargs,
+                    )
+
+                    if isinstance(fx_frontend, frontend.FxFrontendUnpackKwargs):
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:119]: This looks strange. I feel we should try to keep the original input schema in FX exporter instead of making special handling in this class.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                        graph_module, new_args = fx_frontend(model, *args, **kwargs)
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:120]: What if user want to feed another set of args and kwargs? Do they call `graph_module, new_args = fx_frontend(model, *args, **kwargs)` again? Or there is a simpler API such as `new_args, new_kw_args = fx_frontend.normalize_args_kwargs(args, kwargs)`?

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            # Important to rename the test method so that DecorateInfo can find it
+            test_name = base_method.__name__
+
+            # TODO(bowbao): Potentially re-use 'add_decorate_info' to have finer-grained
+            # control over which modules to skip or expect to fail, based on frontend.
+
+            setattr(cls, test_name, base_method)
+
+
+TestFxFrontendConsistency.parameterize_frontends(FX_FRONTENDS)
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:151]: Why do you need to `parameterize_frontends(....)`? How are the elements in FX_FRONTENDS related to the actual test code? I see you dynamically genearte some `def test_*` in `TestFxFrontendConsistency`. But test code should be easier to read and review. For example,
```
class TestFxFrontendConsistency(common_utils.TestCase):
  def test_first_frontend(...):
  def test_second_frontend(...)
```

File: torch/onnx/_internal/fx/frontend.py (hunk 1 above)
Comment (wschin) [reply torch/onnx/_internal/fx/frontend.py:39]: Yes, let's be explicit in code. My first impression to `FxFrontend.__calll__` or `FxFrontend.__calll__` is a question mark. If it's named after what that function really does, it'd be easier to get a quick understanding.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 3 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py:20]: > Can you test this a simple nn.Module or a 1-op function so that aotautograd doesn't alter the function?

I think it only works on models with no parameters.

> This makes me feel we need to have the ability to select what frontend to test for which model.

This is required. More like a decorator of `expect_failure` based on the frontend. It's part of TODO.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 6 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py:93]: I guess you meant a ~~class~~ method of `module_info`?

It's actually an attribute, assigned when initializing `ModuleInfo`.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 7 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py:112]: good catch, this should be controlled by `training`.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 8 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py:119]: You are right, this is not good looking.

Root cause being the assumption of only positional args are allowed by fx graph infrastructures/utilities/usages from almost everywhere. 

This part needs more thoughts. When writing this PR I was leaning towards gradually removing the need for `FxFrontendUnpackKwargs` as we eventually keep original input schema everywhere. But now I think what we need is a builtin input formatter and validator. I want to leave this as TODO since it's a larger chunk of component, to not block the other PRs after this.

File: torch/onnx/_internal/fx/frontend.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        kwargs = copy.deepcopy(kwargs)
+        return self._inner_call(model, *args, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Returns the name of the frontend class formatted with configurations."""
+        return f"{self.__class__.__name__}"
+
+
+class FxFrontendUnpackKwargs:
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/frontend.py:79]: Could you elaborate what important mechanism? And who is `people` referring to? exporter devs or users?

I feel as we move from POC to real product, we need to find the balance between hackability and stable api. If something important has to be hacked, then it means the design is missing something and we should address that.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 4 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py:157]: Added doc in `Note [Add new module test cases]`.

It means the `__init__` of that module takes 0 arguments.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 5 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py]: Cleaned up and added more, in terms of how to add new test cases.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 9 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py]: There is `bind_args_kwargs` for this purpose. But inputs wrangling deserves much more support and attention. I will improve this in follow up PRs.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+                    )
+
+        test_name = f"test_output_match_frontend_{fx_frontend.name}"
+        _output_match_base.__name__ = test_name
+        return _output_match_base
+
+    @classmethod
+    def parameterize_frontends(cls, frontends: Sequence[_FrontendType]):
+        for fx_frontend in frontends:
+            # Generate a test method for each frontend.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py]: Removed this. Previously both frontends and test modules are parameterized on top of `test_output_match` method.

Now only test modules are parameterized on top of `test_output_match` method, via torch utility function `common_modules.modules()`.
frontends are parameterized on top of test class itself. Providing amended class name and attribute `frontend_info`.

The goals are
1. Easy to add new module test cases and new fx frontend, no code copy-paste involved.
2. Module test cases are re-usable elsewhere outside of this test.

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 10 above)
Comment (BowenBao) [reply test/onnx/internal/fx/test_frontend_consistency.py]: Commented above.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+                    pytree.tree_flatten(outputs)[0],
+                    pytree.tree_flatten(fx_outputs)[0],
+                )
+
+
+# 'Parameterize' the test class to run for each fx frontend.
+# Adopted from `parameterized` package.
+# The reason not using `parameterized` package is that it doesn't support
+# applying additional `common_device_type.instantiate_device_type_tests`.
+for frontend_info in FX_FRONTEND_INFOS:
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:273]: I am still not convinced to use code-gen style way to generate tests. We can discuss offline.

**Review by wschin — CHANGES_REQUESTED** [review]
Left a blocking comment about code-gen test. Let's talk.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+        MODULES_DB, train_eval_mode=common_modules.TrainEvalMode.eval_only
+    )
+    def test_output_match(
+        self,
+        device: str,
+        dtype: torch.dtype,
+        module_info: common_modules.ModuleInfo,
+        training: bool,
+    ):
+        """Base test method for testing each module_info with fx frontend."""
```
Comment (wschin) [inline test/onnx/internal/fx/test_frontend_consistency.py:217]: Looks like a key function to create tests with real document?

**Comment by BowenBao** [comment]
@wschin Sure, I have moved this to the end of stack as others are not depending on it.

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+                    pytree.tree_flatten(fx_outputs)[0],
+                )
+
+
+# 'Parameterize' the test class to run for each fx frontend.
+# Adopted from `parameterized` package.
+# The reason not using `parameterized` package is that it doesn't support
+# applying additional `common_device_type.instantiate_device_type_tests`.
+for frontend_info in FX_FRONTEND_INFOS:
+    test_class_module = sys.modules[_TestFxFrontendConsistency.__module__].__dict__
```
Comment (justinchuby) [inline test/onnx/internal/fx/test_frontend_consistency.py:274]: I improved the parameterization in https://github.com/pytorch/pytorch/pull/92087/files#diff-e968c9cb6fc6631cab526cb3a9fe66358c4c6e757e2a223a224b976471bcb753R403-R412. May worth taking a look

File: test/onnx/internal/fx/test_frontend_consistency.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+import torch
+from torch import nn
+from torch.onnx._internal.fx import frontend
+from torch.testing._internal import common_device_type, common_modules, common_utils
+from torch.utils import _pytree as pytree
+
+_FrontendType = Union[frontend.FxFrontend, frontend.FxFrontendUnpackKwargs]
+
+
+class FrontendInfo:
```
Comment (justinchuby) [inline test/onnx/internal/fx/test_frontend_consistency.py:52]: Curious if dataclass wasn't good for this?

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 13 above)
Comment (justinchuby) [reply test/onnx/internal/fx/test_frontend_consistency.py:273]: If the test body is common like we have now, I think it's worth parameterizing. Alternatively we _could_ extract the test body into a function, create multiple modules that call this function etc, but that would mean either (1) two growing list to maintain (FrontendInfo and the test classes) or (2) scattered config (no FronendInfo, skips in every method)

File: test/onnx/internal/fx/test_frontend_consistency.py (hunk 15 above)
Comment (wschin) [reply test/onnx/internal/fx/test_frontend_consistency.py:275]: `parameterization` is not a good idea in Python `unittest`. It increases debug cost and reduces readability without providing actual benefits. If there is something you want to parameterize, you can use those parameters as function arguments.

Comment (justinchuby) [reply test/onnx/internal/fx/test_frontend_consistency.py:275]: In theory we could parameterize the individual test methods. Is that what you mean by using parameters as function arguments? It’s harder to make it work here because PyTorch already has their own way of parameterizing OpInfos. We _could_ also extend PyTorch’s test mechanism with additional options we use, but that can mean effort. 

Can you say more on why it increases debug cost?

**Review by thiagocrepaldi — COMMENTED** [review]
Is this PR still applicable after the design rounds we have been working? This seems to be similar to the `FXGraphExtractor` we have merged

**Comment by BowenBao** [comment]
@thiagocrepaldi no it's not applicable. I sort of linked this to your extractor PR last time calling out this idea (fxfrontend/extractor) is finally in. This code has a poc for aot though, feel free to check out.
